cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -175,3 +175,40 @@ def dephase(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
     return transformer_primitives.map_operations(
         circuit, dephase, deep=True, tags_to_ignore=ignored
     ).unfreeze()
+
+
+@transformer_api.transformer
+def drop_terminal_measurements(
+    circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
+) -> 'cirq.Circuit':
+    """Removes terminal measurements from a circuit.
+
+    This transformer is helpful when trying to capture the final state vector
+    of a circuit with many terminal measurements, as simulating the circuit
+    with those measurements in place would otherwise collapse the final state.
+
+    Args:
+        circuit: The circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options
+            for transformers.
+    Returns:
+        A copy of the circuit, with identity or X gates in place of terminal
+        measurements.
+    Raises:
+        ValueError: if the circuit contains non-terminal measurements.
+    """
+
+    def flip_inversion(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if isinstance(op.gate, ops.MeasurementGate):
+            return [
+                ops.X(q) if b else ops.I(q) for q, b in zip(op.qubits, op.gate.full_invert_mask())
+            ]
+        return op
+
+    if not circuit.are_all_measurements_terminal():
+        raise ValueError('Circuit contains a non-terminal measurement.')
+
+    ignored = () if context is None else context.tags_to_ignore
+    return transformer_primitives.map_operations(
+        circuit, flip_inversion, deep=True, tags_to_ignore=ignored
```

Comment:
I'm following the pattern of `dephase_measurements` above, which uses `deep=True` - should that also change?

There's a complication here in that the meaning of "terminal measurements" falls apart if you don't recurse into subcircuits. Also, that `are_all_measurements_terminal` _does_ go deep, regardless of the context.

Reply:
> I'm following the pattern of dephase_measurements above, which uses deep=True - should that also change?

I think yes, to maintain consistency with rest of the transformer infrastructure. 

Maybe we can update the default value of the `context` parameter to be:

```python
context: Optional['cirq.TransformerContext'] = cirq.TransformerContext(deep=True)
```

instead of None? And then, if we believe that the transformer only makes sense when `deep=True`, then we can raise an error if `context is None or context.deep is False` ? 

This way, we will make sure that a user is not able to do:
```python
>>> circuit = cirq.drop_terminal_measurements(circuit, context=cirq.TransformerContext(deep=False))
```
and get back a result where measurements inside sub-circuits were also dropped ?

Reply:
Done and done. Only `drop_terminal_measurements` has the error case, since non-deep dephasing is still meaningful.